x/curating/spec/README.md

```diff
@@ -0,0 +1,230 @@
+
+- A mapping from `0x01|vendorID|postID` to `Post`.
+- A mapping from `0x02|vendorID|postID|curator` to `Upvote`.
+
+### Post Curation Queue
+
+All queues objects are sorted by timestamp. The time used within any queue is
+first rounded to the nearest nanosecond then sorted. The sortable time format
+used is a slight modification of the RFC3339Nano and uses the the format string
+`"2006-01-02T15:04:05.000000000"`. Notably this format:
+
+- right pads all zeros
+- drops the time zone info (uses UTC)
+
+In all cases, the stored timestamp represents the maturation time of the queue
+element.
+
+For the purpose of tracking the end of post curation windows the post curation queue is kept.
+
+- PostCurationQueue: `0x41 | format(CurationEndTime) -> []VPPair`
+
+```go
+type VPPair struct {
+    VendorID uint32
+    PostID   string
+}
+```
+
+During each `EndBlock`, all the posts that have reached the end of their curation window are processed. The upvotes are tallied, and rewards are paid out to via a mix of quadratic voting and quadratic finance matching.
+
+## Messages
+
+### MsgPost
+
+A social media post to be curated is registered on-chain with `MsgPost`. Note that the actual data of the post is not stored, just a hash of the data, and a vendor and post identifier.
+
+Post curation begins as soon as the post is registered on-chain.
+
+```go
+type MsgPost struct {
+	VendorID      uint32 // 1 = twitter, 2 = reddit, etc.
+	PostID        string
+	Creator       sdk.AccAddress
+	RewardAccount sdk.AccAddress // optional
+	Body          string // hex string, optional
+	Deposit       sdk.Coin
+}
+```
+
+Posts are persisted with the key `0x01|vendorID|postID`.
+
+An optional reward account enables rewards to go to a different account than the creator. 
+
+Validate the post, calculate the curation end time, and insert the vendor/post_id pair into the post curation queue.
+
+### MsgUpvote
+
+Every transaction on Stakebird includes a deposit which could be slashed by governance for bad behavior. This helps keep the network honest.
+
+The amount of vote credits (`VoteAmount`) is quadratically associated to `VoteNum`, and is used for the upvote amount.
+
+| `VoteNum` | `VoteAmount`  |
+|-----------|---------------|
+|   2       | 4             |
+|   3       | 9             |
+|   4       | 16            |
+
+
+```go
+type MsgUpvote struct {
+	VendorID uint32
+	PostID   string
+	Curator  sdk.AccAddress
+	VoteNum  uint32 // 1 = 1, 2 = 4, 3 = 9 (quadratic)
+	Deposit  sdk.Coin
+}
+```
+
+Upvotes are persisted with the key `0x02|vendorID|postID|curator`.
+
+If the corresponding post doesn't exist yet for an upvote, then also create the post with the curator being the `RewardAccount` for the creator. This way, the curator earns creator rewards as well.
+
+### MsgModerate
+
+`MsgModerate` is to signify spam or illegal content. Moderation delegates to the Cosmos governance module, creating a new text proposal for each moderation message.
```

Comment:
Where is this text proposal stored? I don't see it in the `MsgModerate` struct.

I imagine if I click a future moderate button, I'll be able to enter some text explaining why I think it should be moderated as well as specify a deposit... is that correct?

Reply:
Good question. I'm designing this currently as as wrapper around the Cosmos Governance module. So a moderate message will wrap a Cosmos text proposal which I was going to fill with something like:
```
title: "content moderation proposal {vendor_id} - {post_id}"
description: hash of the post
type: "text"
deposit: same deposit from `MsgModerate`
```
Then it will be governance's job to find the content on the relevant network and match it against the hash. I'll add this to the spec :)

Reply:
This won't handle the redistribution of rewards/slashing though, so created https://github.com/public-awesome/stakebird/issues/51 for that.

Reply:
Nice!